plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -443,13 +444,40 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          schedule_production_loop();
       }
 
-      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, next_function<transaction_trace_ptr> next) {
+      void on_incoming_transaction_sync(const packed_transaction_ptr &trx, bool persist_until_expired, bool read_only, next_function<transaction_trace_ptr> next) {
+         chain::controller& chain = chain_plug->chain();
+         const auto max_trx_time_ms = _max_transaction_time_ms.load();
+         fc::microseconds max_trx_cpu_usage = max_trx_time_ms < 0 ? fc::microseconds::maximum() : fc::milliseconds( max_trx_time_ms );
+
+         auto result = transaction_metadata::create_no_recover_keys(trx, transaction_metadata::trx_type::read_only);
+
+         try {
+            if( !process_incoming_transaction_async( result, persist_until_expired, next ) ) {
+               if( _pending_block_mode == pending_block_mode::producing ) {
+                  schedule_maybe_produce_block( true );
+               } else {
+                  restart_speculative_block();
+               }
+            }
+         } catch ( fc::exception_ptr ex) {
+            fc_dlog(_trx_failed_trace_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${txid} : ${why} ",
+                    ("txid", trx->id())("why",ex->what()));
+            next(ex);
+
+            if (_trx_trace_failure_log.is_enabled(fc::log_level::debug)) {
+               auto entire_trx = chain_plug->get_entire_trx(trx->get_transaction());
+               fc_dlog(_trx_trace_failure_log, "[TRX_TRACE] Speculative execution is REJECTING tx: ${entire_trx}", ("entire_trx", entire_trx));
+            }
+         }
+      }
+
+      void on_incoming_transaction_async(const packed_transaction_ptr& trx, bool persist_until_expired, bool read_only, next_function<transaction_trace_ptr> next) {
```

Comment:
Doesn't look like read_only arg is used.

Reply:
Now properly used in constructing transaction metadata.